core/amber/src/main/python/pytexera/udf/examples/test_count_batch_operator.py

```diff
@@ -0,0 +1,83 @@
+    def count_batch_operator(self):
+        return CountBatchOperator()
+
+    def test_count_batch_operator(self, count_batch_operator):
+        count_batch_operator.open()
+        for i in range(27):
+            deque(count_batch_operator.process_tuple(Tuple({"test-1": "hello", "test-2": 10}), 0))
+        deque(count_batch_operator.on_finish(0))
+        batch_counter = count_batch_operator.count
+        assert batch_counter == 3
+        count_batch_operator.close()
+
+    def test_count_batch_operator_simple(self, count_batch_operator):
+        count_batch_operator.open()
+        for i in range(20):
+            deque(count_batch_operator.process_tuple(Tuple({"test-1": "hello", "test-2": 10}), 0))
+        deque(count_batch_operator.on_finish(0))
+        batch_counter = count_batch_operator.count
+        assert batch_counter == 2
+        count_batch_operator.close()
+
+    def test_count_batch_operator_medium(self, count_batch_operator):
+        count_batch_operator.open()
+        for i in range(27):
+            deque(count_batch_operator.process_tuple(Tuple({"test-1": "hello", "test-2": 10}), 0))
+        deque(count_batch_operator.on_finish(0))
+        batch_counter = count_batch_operator.count
+        assert batch_counter == 3
+        count_batch_operator.close()
+
+    def test_count_batch_operator_hard(self, count_batch_operator):
+        count_batch_operator.open()
+        count_batch_operator.BATCH_SIZE = 10
+        for i in range(27):
+            deque(count_batch_operator.process_tuple(Tuple({"test-1": "hello", "test-2": 10}), 0))
+        count_batch_operator.BATCH_SIZE = 5
+        for i in range(27):
+            deque(count_batch_operator.process_tuple(Tuple({"test-1": "hello", "test-2": 10}), 0))
+        deque(count_batch_operator.on_finish(0))
+        batch_counter = count_batch_operator.count
+        assert batch_counter == 9
+        count_batch_operator.close()
+
+    def test_edge_case_string(self, count_batch_operator):
+        count_batch_operator.open()
+        count_batch_operator.BATCH_SIZE = "test"
+        for i in range(55):
+            deque(count_batch_operator.process_tuple(Tuple({"test-1": "hello", "test-2": 10}), 0))
+        deque(count_batch_operator.on_finish(0))
+        batch_counter = count_batch_operator.count
+        assert batch_counter == 2
+        count_batch_operator.close()
+
+    def test_edge_case_non_positive(self, count_batch_operator):
+        count_batch_operator.open()
+        count_batch_operator.BATCH_SIZE = 0
+        for i in range(55):
+            deque(count_batch_operator.process_tuple(Tuple({"test-1": "hello", "test-2": 10}), 0))
+        deque(count_batch_operator.on_finish(0))
+        batch_counter = count_batch_operator.count
+        assert batch_counter == 2
+        count_batch_operator.close()
+
+    def test_edge_case_none(self, count_batch_operator):
+        count_batch_operator.open()
+        count_batch_operator.BATCH_SIZE = None
```

Comment:
just a thinking, what happens if a batch size is changed to None during execution? do we allow or disallow it?

Reply:
My suggestion is if we want to allow this, maybe the batch size is none and we can treat it as returning the whole table. If we don't allow it, maybe send the message to suggest table operator.

Reply:
then what is the difference between batch API and table API when batch size is None? I suggest we explicitly separate them and treat `None` as an invalid batch size. in earlier lines I remembered you commented "batch size has to be positive int", which does not include None.

Reply:
I think this is a good idea. I will use this one. Is what I should do is raising error?

Reply:
you decide. we need to handle edge cases in a reasonable way.